### ISSUE TITLE
Restoring and Using Optical Flow Modules

### ISSUE THREAD

**Issue by AziziShekoofeh** [tensorpack/tensorpack#1249]
Hey, thanks for the great package.

I am trying to import a pretrained model (optical flow) from tensorpack inside my current graph which is a pure tensorflow. The underlying behavior of my graph is temporal which encounter the repeat of optical flow module for each time steps. So, I have followed your suggestions to define and build the graph. And, since I want to share/reuse weights I have also defined argscopes and also tf.REUSE.

When I am starting my session, I am trying to load the pretrained model to my current session. Here, I also followed you suggestion to restore weights. 

```
restorer = get_model_loader(filename) 
# restorer._prms as DicRestor Object
restorer._run_init(sess=sess)
```
However, as I have the 'argscope', the variable names of graph and the pretrained model is not the same (they have the standard prefix). There are a few ways in this case: (1) one way is to rename the keys in dictionary, and (2) the other way is to assign the dic values to the graph variables. 

```
# Method 1
def load_flownet_model(model_path, session):
    # restore the checkpoint for optical flow pretrained model
    restorer = get_model_loader(model_path)
    assert isinstance(restorer, DictRestore)
    model_dict = restorer._prms

    # write out all of the current params
    cnt = 0
    list_var = []
    for name, _ in model_dict.items():
        print(name)
        list_var.append(name)
        cnt += 1
    print("The total number of %d variable has been retrieved from the model!" % cnt)

    cnt = 0
    train_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES)
    for v in train_vars:
        name = v.name
        tshape = v.get_shape()
        # numpy from here
        value = v.eval()
        shape = value.shape
        print(name, tshape, shape)
        cnt += 1
    print("The total number of %d variable has been retrieved from the graph!" % cnt)

    [renameKey(model_dict, name, 'optical_flow_layer/flownet_cell/' + name) for name in list_var]
    # write out all of the current params
    cnt = 0
    for name, _ in model_dict.items():
        print(name)
        cnt += 1
    print("The total number of %d variable has been retrieved from the model!" % cnt)

    # model_dict['optical_flow_layer/flownet_cell/' + name] = model_dict.pop(name)
    restorer._run_init(sess=session)


def renameKey(iterable, old_key, new_key):
    if type(iterable) is dict:
        for key in iterable.keys():
            if key == old_key:
                iterable[new_key] = iterable.pop(key)
                print(new_key)
                return
    else:
        print("not a dictionary")

```

```
# Method 2
def get_variable_to_restore(model_path, session):
    # restore the checkpoint for optical flow pretrained model
    restorer = get_model_loader(model_path)
    assert isinstance(restorer, DictRestore)
    model_dict = restorer._prms

    # write out all of the current params
    cnt = 0
    list_var = []
    for name, _ in model_dict.items():
        print(name)
        list_var.append(name)
        cnt += 1
    print("The total number of %d variable has been retrieved from the model!" % cnt)
    [renameKey(model_dict, name, 'optical_flow_layer/flownet_cell/' + name) for name in list_var]

    cnt = 0
    train_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES)
    assign_ops = []
    for v in train_vars:
        name = v.name
        tshape = v.get_shape()
        # numpy from here
        value = v.eval()
        shape = value.shape
        print(name, tshape, shape)
        if model_dict[name] is not None:
            assign_ops.append(tf.assign(v, tf.cast(model_dict[name].astype(np.float32), v.dtype)))
        cnt += 1
    session.run(assign_ops)
    print("The total number of %d variable has been retrieved and assigned to the graph!" % cnt)
```

Unfortunately, none of them giving me the correct values for the test images. I have the exact same preprocessing step in my code, too. So I have run a few simple test, both using your test code (flownet2.py) and my restore model to compare. Resulting in these two questions:

(1) Interestingly, using the pre-trained models and your flownet2.py code, the optical flow of two still frames (two identical image, or two totally black or white images) are not near zero and they generate some patterns with different and random color values. My question is that if this much of errors is normal? 

(2) And also, do you have a more straight forward suggestion, either for variable key changes in the restored dictionary or eliminating the argscope names to make sure the reproducing of the results in the restored model? 

any recommendation or suggestion could be helpful. Many thanks.

**Comment by ppwwyyxx** [comment]
> (1) Interestingly, using the pre-trained models and your flownet2.py code, the optical flow of two still frames (two identical image, or two totally black or white images) are not near zero and they generate some patterns with different and random color values. My question is that if this much of errors is normal?

I can verify this issue, but running on two consecutive frames works well.
This looks more like an issue with the visualization logic. For example, this line normalizes the color with a data-dependent normalizer which looks suspicious: https://github.com/tensorpack/tensorpack/blob/ca0969089847c37a893a8e99317214c5899278db/examples/OpticalFlow/helper.py#L100
@PatWie do you have cycles to take a look at why this happens? I'm not very familiar with the flow code.

**Comment by ppwwyyxx** [comment]
> (2) And also, do you have a more straight forward suggestion, either for variable key changes in the restored dictionary or eliminating the argscope names to make sure the reproducing of the results in the restored model?

`np.load` will give you the dict and you can do whatever renaming on that dict, and set them to the variables in your graph. This is basically what you did in Method 2, and on a high level the strategy should definitely work.

**Comment by AziziShekoofeh** [comment]
Thank you so much for your prompt response. 

(1) I thought maybe for any reproduction matter I should provide my current version of the software. My tensorpack version is "0.9.5", my Python version is "Python 3.5.2 ", and also my Tensorflow version is "1.11.0"

(2) This is what I did to minimally change the flownet2.py code for evaluation purpose:
- Test 1: inside apply() function, I added to following line after imread(): 
```
    left = np.zeros_like(left)
    right = np.zeros_like(right)
``` 
![flow output_screenshot_02 07 2019 - flownet2-c](https://user-images.githubusercontent.com/8812771/60546942-520b1000-9cd3-11e9-8c28-48169db111e5.png)
![flow output_screenshot_02 07 2019-flownet2](https://user-images.githubusercontent.com/8812771/60546944-52a3a680-9cd3-11e9-95f0-d053fa91ab35.png)

- Test 2: load the same image as the left and right image. I used DAVIS/Bear/00000.jpg

![flow output_screenshot_02 07 2019-flownet2-bear](https://user-images.githubusercontent.com/8812771/60546955-58998780-9cd3-11e9-9db8-e358d529e6b0.png)
![flow output_screenshot_02 07 2019-flownet2-c-bear](https://user-images.githubusercontent.com/8812771/60546957-59321e00-9cd3-11e9-9ff1-5276313d3c3a.png)

The above results are for FlowNet2 and FlowNet2-C networks, respectively. Interestingly, in the FlowNet2 cases the edges of the bear body and background scene are distinguishable.  However, the results seems reasonable when you actually use two distinguishable and non-identical images.

**Comment by PatWie** [comment]
@ppwwyyxx is correct. You cannot expect these optical flow fields to be blank with the current implementation. Any small change (noise) will be amplified in
https://github.com/tensorpack/tensorpack/blob/ca0969089847c37a893a8e99317214c5899278db/examples/OpticalFlow/helper.py#L86

And there are some numerical differences between the original Caffe model and the re-implementation (within differences within 1e-7). 

Have you tried the reference implementation? I haven't tested the corner cases there. It might also produce these artifacts.

The pattern in your post looks clearly like noise in the first optical flow estimate which is then simply upscaled.

**Comment by ppwwyyxx** [comment]
@PatWie thanks for the confirmation. Asking because I'm not familiar with the physical meaning of this normalizer -- is it OK to use a constant normalizer? This may also make visualizations more consistent between a sequence of (>2) frames. Right now when I tried on >2 frames, the outputs of each pair have dramatically different background colors.

**Comment by AziziShekoofeh** [comment]
@PatWie and @ppwwyyxx Thanks for the clarification and the prompt response. The above results are from the original code and the models that you provided inside /examples/OpticalFlow/. I am going to try this corner cases with the other re-implementation (PyTorch, and so on) to see how they perform, if I get any clue I will let you know.

**Comment by PatWie** [comment]
By all means I am not an expert in optical flow either.

But the visualization seems really just showing what is not there. As a quick test you might want to just print the values to the terminal. Just to confirm they are indeed in the same range as in FlowNet2-C. In would be interesting to see what's the mean, max and min of the optical flow values.

Looking at different implementations of flow visualizations most keep max_rad as a user specified parameter btw.

Could you attach the individual input image here (drag and drop). Otherwise would be difficult to reproduce.

**Comment by ppwwyyxx** [comment]
Thanks! It looks like `max_rad = max(max_rad, 1.0)` is good enough in filtering out the noises. This is implemented in the above commit. Now the visualization for identical frames should be empty.
A typical `max_rad` is 10~40 in several videos I tested. Setting it to a constant 40 also seems to work well.

**Comment by AziziShekoofeh** [comment]
@ppwwyyxx @PatWie Thanks :)

**Comment by PatWie** [comment]
@ppwwyyxx The maximum displacement for the correlation layer is fixed
https://github.com/tensorpack/tensorpack/blob/7e7e577fe301302632d0232967c39f26d2908849/examples/OpticalFlow/flownet_models.py#L401
As there are several transposed convolutions afterwards, the `max_rad` might become much larger than 40 in some corner cases. Your current fix seems to be better.
